R/getInsertSizeDistFromBam.R

```diff
@@ -0,0 +1,108 @@
+#'   tabulated together.
+#' @param regions \code{GRanges} object. Only alignments falling into these
+#'   regions will be used. If \code{NULL} (the default), all alignments are
+#'   used.
+#' @param nmax \code{numeric(1)} specifying the maximal number of alignments to
+#'   read. If \code{NA} (the default), the alignments in \code{regions} (if
+#'   \code{regions} are not \code{NULL}) or in the bam file will be used.
+#' @param isizemax \code{numeric(1)} specifying the maximal insert size to
+#'   report. Larger insert sizes will be set to \code{isizemax} with a warning.
+#'
+#' @return \code{integer} vector with the number of insert sizes. The element at
+#'   position \code{i} gives the observed number of alignment pairs with an
+#'   insert size of \code{i}. The number of insert sizes greater than
+#'   \code{isizemax} that were set to \code{isizemax} are reported in the
+#'   attribute \code{"ncapped"}.
+#'
+#' @seealso \code{\link[Rsamtools]{scanBam}} used to read alignments.
+#'
+#' @examples
+#' if (requireNamespace("Rsamtools", quietly = TRUE)) {
+#'     bamf <- system.file("extdata", "getInsertSizeDistFromBam", "atac_mm10.bam",
+#'                         package = "swissknife")
+#'     isize <- getInsertSizeDistFromBam(bamf)
+#'     attr(isize, "ncapped")
+#'     plot(isize, type = "l",
+#'          xlab = "Insert size (bp)", ylab = "Number of fragments")
+#' }
+#'
+#' @importFrom IRanges IRangesList
+#' 
+#' @export
+getInsertSizeDistFromBam <- function(fname,
+                                     regions = NULL,
+                                     nmax = NA_integer_,
+                                     isizemax = 800) {
+    .assertPackagesAvailable(c("Rsamtools"))
+    if (!is.null(regions) && !inherits(regions, "GRanges")) {
+        stop("'regions' must be either NULL or a GRanges object.")
+    }
+    if (!is.na(nmax) && (!is.numeric(nmax) || length(nmax) != 1L || nmax <= 0)) {
+        stop("'nmax' must be either NA or a numeric scalar greater than zero.")
+    }
+    if (!is.numeric(isizemax) || length(isizemax) != 1L || isizemax <= 0) {
+        stop("'isizemax' must be a numeric scalar greater than zero.")
+    }
+    
+    # set scanning parameters
+    if (is.null(regions)) {
+        regions <- IRanges::IRangesList()
+    }
+    params <- Rsamtools::ScanBamParam(
+        flag = Rsamtools::scanBamFlag(isPaired = TRUE,
+                                      isProperPair = TRUE,
+                                      isUnmappedQuery = FALSE,
+                                      hasUnmappedMate = FALSE,
+                                      isFirstMateRead = TRUE,
+                                      isSecondaryAlignment = FALSE,
+                                      isNotPassingQualityControls = NA,
+                                      isDuplicate = FALSE,
+                                      isSupplementaryAlignment = FALSE),
+        what = c("isize"),
```

Comment:
are there any applications where the user may want to vary the parameters for scanBam? I am thikning of PE non-ATAC data (for example PE RNA-seq/ChIP-seq/HiC), and if we can generalize a function to get iSIze, but have the defaults be specific to ATAC since that is where we use it most often in practice.

Reply:
Initially I also thought of making `params` an argument of the function. I decided against it for two reasons: 
1. `Rsamtools` is in `Suggests` to speed up package load time, so I cannot put `Rsamtools::ScanBamParam(...)` as a default value into the formals of the function (one could of course set the default to zero and set it internally after the check if `Rsamtools` is available, but that would hide the default value as much as now.
2. More importantly, I could not think of a use-case that would require different parameters, i.e. I could not think of a situation, where I would want to keep secondary alignments and get multiple `isize` values for the same pair of reads.
I am happy to change this is you think it would be better to be able to change `params`.

Reply:
I see, yes I was thinking outloud. The way it is implemented now is really how I would use it most commonly when I want to look at these plots. The one parameter I don't have a good intuition about is `isDuplicate = FALSE` since in ATAC-seq maybe we can have more reads that appear as duplicates for short fragments, since the enzyme cutting site is specific. But I think either way wouldn't affect the final plot so much. So I would leave it as is :)

Reply:
Ok, I'll leave it as it is then for the moment. If we find a use-case that needs different `params`, we can always expose it later. The `isDuplicate` by the way is usually not set by the aligners we use (it may indicate optical or PCR duplicates, and could for example be set by an aligner that also uses information from UMIs). I think for most of our bam files it will not make a difference, but I set `isDuplicate = FALSE` because I thought we would want to exclude alignments that are clearly flagged as duplicates in general.